lib/logstash/plugin_mixins/event_support/event_target_decorator.rb

```diff
@@ -0,0 +1,116 @@
+require 'logstash/event'
+require 'forwardable'
+
+module LogStash; module PluginMixins; module EventSupport
+
+  # `LogStash::Event` decorator which supports most of the event API.
+  # Values end up being stored in a (prefix hash) target namespace.
+  #
+  # @note Internal references (starting with `@`) e.g. `@timestamp` or
+  # `@metadata` are respected and thus are NOT namespaced.
+  class EventTargetDecorator
+    extend Forwardable
+
+    def self.wrap(event, target_namespace = nil)
+      self.new(event, target_namespace)
+    end
+
+    def initialize(event, target_namespace)
+      @event = event
+      namespace = target_namespace.to_s.strip
+      if namespace.empty? || namespace.start_with?('[')
+        @target_ns = namespace
+      else
+        @target_ns = "[#{namespace}]"
+      end
+    end
+
+    # Event interface
+
+    def_delegators :@event, :cancel, :uncancel, :cancelled?, :clone # no args
+    def_delegators :@event, :to_s, :to_hash, :to_hash_with_metadata, :to_json
+    def_delegators :@event, :tag, :timestamp, :timestamp=
+
+    # @return value at [target_namespace][ref]
+    #
+    # @note Internal references @timestamp and @metadata are NOT namespaced.
+    def get(ref)
+      @event.get(target_ref(ref))
+    end
+
+    # @return set value
+    #
+    # @note Internal references @timestamp and @metadata are NOT namespaced.
+    def set(ref, value)
+      @event.set(target_ref(ref), value)
+    end
+
+    # @return whether reference is present in event
+    #
+    # @note Internal references @timestamp and @metadata are NOT namespaced.
+    def include?(ref)
+      @event.include?(target_ref(ref))
+    end
+
+    # Removes a mapping from the event.
+    #
+    # @note Internal references @timestamp and @metadata are NOT namespaced.
+    def remove(ref)
+      @event.remove(target_ref(ref))
+    end
+
+    # Appends data from given event to this event.
+    # @return self
+    def append(event)
+      if @target_ns.empty?
+        namespaced_event = event
+      else
+        namespaced_event = LogStash::Event.new
+        namespaced_event.set @target_ns, event.to_hash
+      end
+      @event.append(namespaced_event) # only data is appended
+      self
+    end
+
+    # Overwrite event from another.
+    # @return self
+    #
+    # @note In this case target namespace is not maintained.
+    def overwrite(event)
+      # TODO unclear whether we really need this to retain namespace when overwriting, might be confusing?!?
```

Comment:
this piece could work retaining namespace but than I was a bit hesitant to implement it